### PR TITLE
Build time reduction: AzStdOnDemandReflection

### PR DESCRIPTION
**Pull request by nemerle**
Moved all fully specialized templates from AzStdOnDemandReflection.inl to cpp file. 
This reduces compilation time and produced profile/debug file sizes.

The specializations for string and string_view are only implemented for 'char' type, since
others are not used anywhere.

Extracted `Reflect` method from `ClientAuthAWSCredentials` to a cpp file.

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/RTTI/AzStdOnDemandReflection.inl [hunk 1]
```diff
@@ -69,105 +88,13 @@ namespace AZ
         
         static void Reflect(ReflectContext* context)
         {
-            if (BehaviorContext* behaviorContext = azrtti_cast<BehaviorContext*>(context))
+            constexpr bool is_string = AZStd::is_same_v<Element,char> && AZStd::is_same_v<Traits,AZStd::char_traits<char>>
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/RTTI/AzStdOnDemandReflection.inl:91]: nitpick: missing spaces right after the template argument commas

**Review by lumberyard-employee-dm — APPROVED** [review]

**Comment by lumberyard-employee-dm** [comment]
@nemerle Your build has failed due to failing to pass the Copyright validator: https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-3111/
```
13:56:17  VALIDATION FAILURE SUMMARY
13:56:17  VALIDATOR_FAILED: CopyrightHeaderValidator D:\workspace\o3de\Gems\AWSClientAuth\Code\Source\Authorization\ClientAuthAWSCredentials.cpp::CopyrightHeaderValidator FAILED - Source file missing O3DE copyright header.
13:56:17  VALIDATOR_FAILED: CopyrightHeaderValidator D:\workspace\o3de\Gems\AWSClientAuth\Code\Source\Authorization\ClientAuthAWSCredentials.cpp::CopyrightHeaderValidator FAILED - Source file missing any recognized copyrights.
```

**Comment by nemerle** [comment]
License added to a file in 84ed37b

**Comment by lumberyard-employee-dm** [comment]
The build has succeeded

**Review by lumberyard-employee-dm — APPROVED** [review]

**Comment by nemerle** [comment]
Is there anything I can do to help in merging this PR?

**Comment by lumberyard-employee-dm** [comment]
> 
> 
> Is there anything I can do to help in merging this PR?

@carlitosan Could you look at this PR?

**Review by amznestebanpapp — APPROVED** [review]

File: Gems/AWSClientAuth/Code/Source/Authorization/ClientAuthAWSCredentials.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#include <AzCore/RTTI/BehaviorContext.h>
+#include <AzCore/RTTI/AzStdOnDemandReflection.inl>
+#include <AzCore/Serialization/SerializeContext.h>
+#include <AzCore/Serialization/AZStdContainers.inl>
+
+namespace AWSClientAuth
+{
+void ClientAuthAWSCredentials::Reflect(AZ::ReflectContext *context)
+{
+    auto serializeContext = azrtti_cast<AZ::SerializeContext*>(context);
```
Comment (moraaar) [inline Gems/AWSClientAuth/Code/Source/Authorization/ClientAuthAWSCredentials.cpp:20]: The function `ClientAuthAWSCredentials::Reflect` is nested inside a namespace, it's missing tabulation.

**Review by moraaar — CHANGES_REQUESTED** [review]
Nice refactor! One minor tabulation change.

File: Gems/AWSClientAuth/Code/Source/Authorization/ClientAuthAWSCredentials.cpp (hunk 2 above)
Comment (nemerle) [reply Gems/AWSClientAuth/Code/Source/Authorization/ClientAuthAWSCredentials.cpp]: Reformatted this file in 4911a56

Comment (moraaar) [reply Gems/AWSClientAuth/Code/Source/Authorization/ClientAuthAWSCredentials.cpp]: Thanks @nemerle 
Approved!

**Review by moraaar — APPROVED** [review]

**Review by lsemp3d — APPROVED** [review]

**Comment by nemerle** [comment]
2 of the 3 CI failures are caused by
```
AssetProcessor: Trace::Error  D:/workspace/o3de/Code/Tools/AssetProcessor/native/AssetDatabase/AssetDatabase.cpp(836): 'bool __cdecl AssetProcessor::AssetDatabaseConnection::PostOpenDatabase(void)'
AssetProcessor: The database in the Cache folder appears to be from a NEWER version of Asset Processor than this one.
```

I'll take a look at the last error (linux side) today.

**Comment by lumberyard-employee-dm** [comment]
> 
> 
> 2 of the 3 CI failures are caused by
> 
> ```
> AssetProcessor: Trace::Error  D:/workspace/o3de/Code/Tools/AssetProcessor/native/AssetDatabase/AssetDatabase.cpp(836): 'bool __cdecl AssetProcessor::AssetDatabaseConnection::PostOpenDatabase(void)'
> AssetProcessor: The database in the Cache folder appears to be from a NEWER version of Asset Processor than this one.
> ```
> 
> I'll take a look at the last error (linux side) today.

I'll kick off a clean build for that now.

**Comment by nemerle** [comment]
Linux build fixed, ( take note that I've merged development into this branch to locate the problems )

**Pull request merged**